[User Model] Notification permission and PushSubscription observer updates

jennantilla opened the pull request:
# Description
## One Line Summary
Update the Notification permission observer and PushSubscription observer to be called by `addEventListener` methods.

## Details

### Motivation
Align the Typescript methods to addEventListener established convention.

### Scope

- Added an additional notification event type `permissionChange` to `addEventListener` and `removeEventListener` methods on Notifications namespace, removed old `addPermissionObserver` and `removePermissionObserver` methods. 
-  Replaced `addObserver` and `removeObserver` methods from PushSubscription namespace with `addEventListener` and `removeEventListener` for event type `change`

# Testing
## Manual testing
Manually tested on an Pixel 4 simulator running Android 13 and iPhone 14 running iOS 16.4. Confirmed these new methods successfully add and remove the observers for Notification permission and PushSubscription. 

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [X] Public API changes

# Checklist
## Overview
   - [X] I have filled out all **REQUIRED** sections above
   - [X] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [X] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [X] I have included test coverage for these changes, or explained why they are not needed
   - [X] All automated tests pass, or I explained why that is not possible
   - [X] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [X] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [X] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1532)
<!-- Reviewable:end -->

jennantilla committed:
run Prettier for linting test failure

jennantilla committed:
Migration guide fixes

jennantilla committed:
Update Migration Guide formatting

jennantilla committed:
Update Notification permission observer and PushSubscription observer to be created via addEventListener methods

jennantilla committed:
Update Migration Guide examples with new addEventListener methods for PushSubscription and Notification permission observers

jennantilla committed:
Run Prettier for new linting errors

jkasten2 commented on src/events/EventManager.ts:
```diff
@@ -89,10 +89,20 @@ export default class EventManager {
-  clearEventHandler(eventName: string) {
-    this.eventHandlerArrayMap.delete(eventName);
+  clearEventHandler(eventName: string, handler: any) {
+    let handlerArray = this.eventHandlerArrayMap.get(eventName);
```
Can be `const` since this variable isn't reassigned.

jkasten2 commented on src/events/EventManager.ts:
```diff
@@ -89,10 +89,20 @@ export default class EventManager {
-    this.eventHandlerArrayMap.delete(eventName);
+  clearEventHandler(eventName: string, handler: any) {
+    let handlerArray = this.eventHandlerArrayMap.get(eventName);
+    if (handlerArray) {
```
These `if` statements are getting pretty nested, recommend doing an early return instead for readability.

jennantilla committed:
Update clearEventHandler method with early return

jennantilla committed:
Fix lint error on EventManager class

jennantilla committed:
Update Limitations and Notifications permission event listener code sample

jennantilla committed:
Update iOS click listeners to be invoked by sharedInstance

jennantilla committed:
Remove old Email and SMS observers from iOS Event Emitter.

jennantilla committed:
Update PushSubscription interfaces, properties, and functions

jennantilla committed:
Delete native call to removePushSubscriptionObserver

jennantilla committed:
Update InAppMessages listeners to be handled by EventManager

jennantilla committed:
Update Notification event interfaces and methods

Notification listeners to be handled by EventManager

jennantilla committed:
Update Android Notification and IAM listeners to send entire event

jennantilla committed:
Update EventManager to allow for multiple Notification and IAM listeners

Remove set setEventHandler since there can now be multiple listeners

jennantilla committed:
Fix lint errors

jennantilla committed:
remove clearHandlers public method

Additional nit: clean up unused import

nan-li reviewed (commented):
> Reviewed 6 of 12 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1532)* status: 6 of 12 files reviewed, 10 unresolved discussions (waiting on @emawby, @jennantilla, and @shepherd-l)

___
*[`MIGRATION_GUIDE.md` line 381 at r2](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1532#-N__LDcy0ajbtbbksTT0:-N__LDcy0ajbtbbksTT1:b-cfkf8j) ([raw file](https://github.com/OneSignal/react-native-onesignal/blob/7c0a465a707619150b25c5f1b88223e2909abe85/MIGRATION_GUIDE.md#L381)):*
> ```Markdown
> - Changing app IDs is not supported.
> - Any `User` namespace calls must be invoked **after** initialization. Example: `OneSignal.User.addTag("tag", "2")`
> - In the iOS SDK, the user state is only refreshed from the server when a new session is started (cold start or backgrounded for over 30 seconds) or when the user is logged in. This is by design.
> ```

Also true on Android, may not be documented in the migration guide.
___
*[`android/src/main/java/com/geektime/rnonesignalandroid/RNOneSignal.java` line 361 at r2](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1532#-N_pFDXcALPEClznfLZs:-N_pFDXcALPEClznfLZt:b-kp7zqc) ([raw file](https://github.com/OneSignal/react-native-onesignal/blob/7c0a465a707619150b25c5f1b88223e2909abe85/android/src/main/java/com/geektime/rnonesignalandroid/RNOneSignal.java#L361)):*
> ```Java
>             sendEvent("OneSignal-notificationWillDisplayInForeground",
>                     RNUtils.convertHashMapToWritableMap(
>                         RNUtils.convertNotificationEventToMap(event)));
> ```

convert notification will display event to map (so not to confuse with the notification click event).
___
*[`android/src/main/java/com/geektime/rnonesignalandroid/RNOneSignal.java` line 506 at r2](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1532#-N_pH6XY3SJzCEg1CV8-:-N_pH6XY3SJzCEg1CV80:bk8qy52) ([raw file](https://github.com/OneSignal/react-native-onesignal/blob/7c0a465a707619150b25c5f1b88223e2909abe85/android/src/main/java/com/geektime/rnonesignalandroid/RNOneSignal.java#L506)):*
> ```Java
>             sendEvent("OneSignal-subscriptionChanged",
>                     RNUtils.convertHashMapToWritableMap(
>                             RNUtils.convertOnSubscriptionChangedToMap(pushSubscription)));
> ```

Pass original event with current and previous
___
*[`ios/RCTOneSignal/RCTOneSignal.h` line 8 at r2](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1532#-N_pHUKF5HrHihG7TLmx:-N_pHUKF5HrHihG7TLmy:b-u1jn6k) ([raw file](https://github.com/OneSignal/react-native-onesignal/blob/7c0a465a707619150b25c5f1b88223e2909abe85/ios/RCTOneSignal/RCTOneSignal.h#L8)):*
> ```objc
> #endif
> 
> @interface RCTOneSignal : NSObject <OSPushSubscriptionObserver, OSNotificationPermissionObserver, OSInAppMessageLifecycleListener, OSInAppMessageClickListener, OSNotificationClickListener>
> ```

Notification Foreground Lifecycle once that one is moved to RCTOneSignal
___
*[`ios/RCTOneSignal/RCTOneSignalEventEmitter.m` line 271 at r2](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1532#-N_pJxmnCdjRDY1fsckx:-N_pJxmnCdjRDY1fscky:b-cbbkt4) ([raw file](https://github.com/OneSignal/react-native-onesignal/blob/7c0a465a707619150b25c5f1b88223e2909abe85/ios/RCTOneSignal/RCTOneSignalEventEmitter.m#L271)):*
> ```objc
> 
> RCT_EXPORT_METHOD(addNotificationForegroundLifecycleListener) {
>     [OneSignal.Notifications addForegroundLifecycleListener:self];
> ```

Modify to use [RCTOneSignal sharedInstance] as well
___
*[`ios/RCTOneSignal/RCTOneSignalEventEmitter.m` line 271 at r2](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1532#-N_pKH6m9tHdna3-7c_K:-N_pKH6m9tHdna3-7c_L:bd9d91y) ([raw file](https://github.com/OneSignal/react-native-onesignal/blob/7dc08f70edcedd2b067c55797144c920b8375a46/ios/RCTOneSignal/RCTOneSignalEventEmitter.m#L271)):*
> ```objc
> }
> 
> RCT_EXPORT_METHOD(clearAllNotifications) {
> ```

Did you mean to remove clearAllNotifications?
___
*[`src/events/EventManager.ts` line 35 at r2](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1532#-N_pPEWM90FXnu-RIicU:-N_pPEWM90FXnu-RIicV:b-17kvmc) ([raw file](https://github.com/OneSignal/react-native-onesignal/blob/7c0a465a707619150b25c5f1b88223e2909abe85/src/events/EventManager.ts#L35)):*
> ```XML
>   private RNOneSignal: NativeModule;
>   private oneSignalEventEmitter: NativeEventEmitter;
>   private eventHandlerArrayMap: Map<string, Array<(event: any) => void>>;
> ```

eventListenerArrayMap



_Code quote:_
```XML
eventHandlerArrayMap
```
___
*[`src/events/EventManager.ts` line 75 at r2](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1532#-N_pPUbl2nP-FWblO4pj:-N_pPUbl2nP-FWblO4pk:b8i7att) ([raw file](https://github.com/OneSignal/react-native-onesignal/blob/7c0a465a707619150b25c5f1b88223e2909abe85/src/events/EventManager.ts#L75)):*
> ```XML
>    * @returns void
>    */
>   clearEventHandler(eventName: string, handler: any) {
> ```

removeEventListener to be clearer that we are not clearing all the listeners


<!-- Sent from Reviewable.io -->

jennantilla committed:
Update Event Manager method names

Change clearEventHandler to removeEventListener and EventHandler to EventListener

jennantilla committed:
remove setLaunchURLsInApp

jennantilla committed:
Add back clearAll method to iOS

jennantilla committed:
Update Android notification and push subscription conversion methods naming/routing

jennantilla committed:
Remove user state refresh limitation from Migration Guide

jennantilla committed:
Update import for push subscription changed state and method declarations for notification click

jennantilla committed:
Replace logLevel types from int to enum

jennantilla committed:
Bump iOS native dependency to beta-5

jennantilla committed:
Update InAppMessage lifecycle listeners to send entire event on iOS

jennantilla committed:
Fix linting errors
